src/geography.cpp

```diff
@@ -26,128 +30,137 @@ using namespace spherely;
-    py::buffer_info result_buf = result.request();
-    int *rptr = static_cast<int *>(result_buf.ptr);
-
-    for (py::ssize_t i = 0; i < buf.size; i++) {
-        auto geog_ptr = (*geographies.data(i)).as_geog_ptr();
-        rptr[i] = geog_ptr->num_shapes();
-    }
-
-    return result;
+std::unique_ptr<s2geog::Geography> Geography::clone_geog() const {
+    return clone_s2geography(*m_s2geog_ptr, m_geog_type);
 }
 
-py::array_t<PyObjectGeography> create(py::array_t<double> xs, py::array_t<double> ys) {
-    py::buffer_info xbuf = xs.request(), ybuf = ys.request();
-    if (xbuf.ndim != 1 || ybuf.ndim != 1) {
-        throw std::runtime_error("Number of dimensions must be one");
-    }
-    if (xbuf.size != ybuf.size) {
-        throw std::runtime_error("Input shapes must match");
-    }
+Geography Geography::clone() const {
+    auto new_geog_ptr = clone_s2geography(*m_s2geog_ptr, m_geog_type);
 
-    auto result = py::array_t<PyObjectGeography>(xbuf.size);
-    py::buffer_info rbuf = result.request();
+    // skip extract properties
+    auto new_object = Geography();
 
-    double *xptr = static_cast<double *>(xbuf.ptr);
-    double *yptr = static_cast<double *>(ybuf.ptr);
-    py::object *rptr = static_cast<py::object *>(rbuf.ptr);
+    new_object.m_s2geog_ptr = std::move(new_geog_ptr);
+    new_object.m_geog_type = m_geog_type;
+    new_object.m_is_empty = m_is_empty;
 
-    for (py::ssize_t i = 0; i < xbuf.shape[0]; i++) {
-        auto point_ptr = PointFactory::FromLatLonDegrees(xptr[i], yptr[i]);
-        // rptr[i] = PyObjectGeography::as_py_object(std::move(point_ptr));
-        rptr[i] = py::cast(std::move(point_ptr));
-    }
+    return new_object;
+}
 
-    return result;
+void Geography::extract_geog_properties() {
+    if (const auto *ptr = downcast_geog<s2geog::PointGeography>(); ptr) {
+        if (ptr->Points().empty()) {
+            m_is_empty = true;
+        }
+        if (ptr->Points().size() <= 1) {
+            m_geog_type = GeographyType::Point;
+        } else {
+            m_geog_type = GeographyType::MultiPoint;
+        }
+    } else if (const auto *ptr = downcast_geog<s2geog::PolylineGeography>(); ptr) {
+        if (ptr->Polylines().empty()) {
+            m_is_empty = true;
+        }
+        if (ptr->Polylines().size() <= 1) {
+            m_geog_type = GeographyType::LineString;
+        } else {
+            m_geog_type = GeographyType::MultiLineString;
+        }
+    } else if (const auto *ptr = downcast_geog<s2geog::PolygonGeography>(); ptr) {
+        const auto &s2poly_ptr = ptr->Polygon();
+        // find the outer shells (loop depth = 0, 2, 4, etc.)
+        std::vector<int> outer_shell_loop_ids;
+
+        outer_shell_loop_ids.reserve(static_cast<size_t>(s2poly_ptr->num_loops()));
+
+        for (int i = 0; i < s2poly_ptr->num_loops(); i++) {
+            if ((s2poly_ptr->loop(i)->depth() % 2) == 0) {
+                outer_shell_loop_ids.push_back(i);
+            }
```

Comment:
Given that you don't really use the `outer_shell_loop_ids` vector, except for checking the size, you could also replace this with a simpler count?

```
int n_outer_shell_loops = 0;
...
for .. 
  if .. {
    n_outer_shell_loops ++;
  }
```